Deal with virtual function calls with no candidate targets

owen-mc-diffblue opened the pull request:
During virtual method resolution, if a virtual function call has no candidate targets then we add the class it was called on as an instantiated class.

Also tidied up the code in various ways.

Commit:
Replace do-while loop with equally valid while loop

It's slightly easier to read a while loop, and it makes sense to
initialise any_new_methods to true because we have already put things
into method_worklist2.

Commit:
Rename two variables and make one more local

Commit:
Remove code duplication

Commit:
Store virtual function calls instead of virtual call-sites

Also, use an unordered list instead of a vector so we don't have to
deduplicate

Commit:
Fix order of parameters in function header

Commit:
Deal with virtual function calls with no candidate targets

During virtual method resolution, if a virtual function call has no
candidate targets then we add the class it was called on as an
instantiated class.

Commit:
Remove incorrect comment

tautschnig commented on src/util/irep.h:
```diff
@@ -29,6 +29,8 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <list>
 #else
 #include <map>
+#include <unordered_set>
```
This is in the wrong place for it would not get included when `SUB_IS_LIST` is set.

tautschnig commented on src/util/irep.h:
```diff
@@ -43,6 +45,7 @@ typedef std::string irep_namet;
 // NOLINTNEXTLINE(readability/identifiers)
 typedef string_hash irep_id_hash;
 #endif
+typedef std::unordered_set<irep_idt, irep_id_hash> id_sett;
```
It's probably a good idea to introduce a widely visible typedef, but I think this needs to happen in a separate PR: 1) Various places already define an `id_sett`; 2) those definitions don't always use the same types; 3) the type name should likely make clear that this is an unordered set of IDs.

owen-mc-diffblue replied:
Okay, for this PR I'll just make it a local typedef, but I might try and make a widely visible one in another PR. For the record it's `std::unordered_set<irep_idt, irep_id_hash>` 9 times and `std::set<cpp_idt *>` 2 times. I like that `id_sett` is short, but I agree that you might assume it wasn't unordered.

NathanJPhillips replied:
Here's a PR for you: https://github.com/diffblue/cbmc/pull/2075

tautschnig reviewed (dismissed):
Please reconsider the `id_sett` part as detailed in my comments. I have not looked at the Java part.

owen-mc-diffblue commented:
@smowton Please review

smowton commented on src/java_bytecode/ci_lazy_methods.cpp:
```diff
@@ -122,46 +122,88 @@ bool ci_lazy_methodst::operator()(
+
+    any_new_classes = false;
+
+    // Find virtual callsites with no candidate targets, guess that the class
```
Suggest giving example: if we have `invokevirtual A.f` and we don't believe either `A` or any of its children may exist, assume specifically `A` is somehow instantiated. Note this may result in an abstract class being classified as instantiated, which stands in for some unknown concrete subclass: in this case the called method will be a stub.

smowton reviewed (commented):
Looks good, but needs tests -- at least the minimal example I gave on Slack please:

```
public class test {

  interface factory_intf {
    public intf getintf();
  }

  interface intf {
    public void f();
  }

  public static void main(factory_intf i) { i.getintf().f(); }
}
```

Commit:
Test lazy-loading when there are no candidates

It should decide that `intf` is instantiated even though (a) it doesn't
see a constructor for it and (b) it is an interface. It isn't ideal
instantiating an interface, but it stands in for a class which
implements the interface.

owen-mc-diffblue commented:
@smowton I have added that comment and that test

NathanJPhillips commented on src/java_bytecode/ci_lazy_methods.cpp:
```diff
@@ -122,46 +122,92 @@ bool ci_lazy_methodst::operator()(
+        candidate_target_methods,
+        symbol_table);
+
+      if(candidate_target_methods.empty())
```
I'd invert this with a continue for clarity of control flow.

NathanJPhillips commented on src/java_bytecode/ci_lazy_methods.cpp:
```diff
@@ -122,46 +122,92 @@ bool ci_lazy_methodst::operator()(
+
+      if(candidate_target_methods.empty())
+      {
+        any_new_classes = true;
```
I'd set this at the end after you've done something that adds new classes.

Commit:
Use unordered set of irep_ids in ci_lazy_methods

Used set instead of vector to remove duplicates, which should prevent
investigating the same method multiple times.

owen-mc-diffblue commented:
@NathanJPhillips I've addressed your comments. I've also removed the `typedef` for `id_sett` and used `std::unordered_set<irep_idt>` instead, following #2082.

smowton commented:
@owen-jones-diffblue test-gen bump?

owen-mc-diffblue commented:
https://github.com/diffblue/test-gen/pull/1710
